### ISSUE TITLE
Missing "alt" release event, OR Why modifier key state should be part of keyboard input event

### ISSUE THREAD

**Issue by jwilm** [tomaka/glutin#802]
I've run into an issue not having modifier keys with the keyboard input events.

My Glutin application tracks modifier states as you might expect. It has several key bindings relying on the state of the _alt_ modifier. When alt-tabbing out of the application, the `Release` event does not fire. The event is not delivered on focus, either. This means my modifier state becomes corrupted because the key is no longer pressed, but my state insists that it is.

Here's the log when I alt-tab out of the application; note the missing Released event:

```
update input state; state=Pressed; key=ALT_LEFT
event: Focused(false)
event: Focused(true)
event: Focused(false)
event: Focused(true)
event: Focused(false)
```

My initial solution to this was to clear that state on the `Focus` event, but that's no good either because I get `Focus(false)` and `Focus(true)` immediately after pressing _alt_. Here's the logs demonstrating that behavior (press and release alt key, didn't actually switch away from window):

```
update input state; state=Pressed; key=ALT_LEFT
event: Focused(false)
event: Focused(true)
update input state; state=Released; key=ALT_LEFT
```

The event not being delivered feels like a bug, but this wouldn't be an issue anyway if the modifier state was still part of the keyboard input event. I've also thought that not passing modifier state with the keyboard event is forcing almost every glutin application to reimplement the same state tracking code themselves. It might be nice if the modifier state were still part of that event (I see it was removed in #92, but there's no discussion as to why).

I'm on Ubuntu 14.04 with X11.

**Comment by jwilm** [comment]
I've looked into this a bit, and I apparently misunderstood the change in #92. Glutin before that PR didn't use the modifier state from the windowing API as I thought. Instead, Glutin tracked key presses internally as a Glutin application must do now.

I reviewed the APIs that Glutin supports. Using the WM API modifier tracking seems feasible.

* x11: `XKeyPress` has a `state` field
* cocoa: `NSEvent` has a `modifierFlags` field
* win32: `GetKeyState` can be called to get all keyboard state at time of event
* wayland: This one is confusing me a bit. It looks like the protocol doesn't include modifier keys with regular key press events, but libraries like `xkbcommon` handle it somehow and provide a modifiers field, so it should be possible. It looks like the Rust wayland wrapper [delivers the modifiers and input events](https://doc.servo.org/wayland_client/sys/wayland/client/enum.WlKeyboardEvent.html). I guess glutin could just track these modifiers internally. Maybe someone more familiar with Wayland could comment?

I'm proposing to add the `Modifiers` field to the `Event::KeyboardInput` variant, but instead of tracking modifier keys internally, use the WM APIs to do so as much as possible (with the exception of wayland, apparently).

If this feature were implemented, would the project accept it? Is this a reasonable approach?
